dirsrvtests/tests/suites/ldapi/ldapi_test.py

```diff
@@ -0,0 +1,114 @@
+
+
+def test_ldapi_authdn_attr_rewrite(topo, request):
+    """Test LDAPI Authentication DN mapping feature
+
+    :id: e8d68979-4b3d-4e2d-89ed-f9bad827718c
+    :setup: Standalone Instance
+    :steps:
+        1. Set LDAPI configuration
+        2. Create LDAP user
+        3. Create OS user
+        4. Create entries under cn=config for auto bind subtree and mapping entry
+        5. Do an LDAPI ldapsearch as the OS user
+        6. OS user was mapped expected LDAP entry
+    :expectedresults:
+        1. Success
+        2. Success
+        3. Success
+        4. Success
+        5. Success
+        6. Success
+    """
+
+    LINUX_USER = "ldapi_test_lib389_user"
+    LINUX_PWD = "5ecret_137"
+    AUTH_DN_ATTR = "authenticateAsDN"
+    LDAP_ENTRY_DN = "uid=test_ldapi,ou=people,dc=example,dc=com"
+
+    def fin():
+        # Remove the OS user
+        try:
+            subprocess.run(['userdel', LINUX_USER])
+        except:
+            pass
+    request.addfinalizer(fin)
+
+    # Must be root
+    if os.geteuid() != 0:
+        return
+
+    # Perform config tasks
+    topo.standalone.config.set('nsslapd-accesslog-logbuffering', 'off')
+    topo.standalone.config.set('nsslapd-ldapiautoauthdnattr', AUTH_DN_ATTR)
+    topo.standalone.config.set('nsslapd-ldapientrysearchbase', 'cn=auto_bind,cn=config')
+    topo.standalone.config.set('nsslapd-ldapimaptoentries', 'on')
+    topo.standalone.config.set('nsslapd-ldapiuidnumbertype', 'uidNumber')
+    topo.standalone.config.set('nsslapd-ldapigidnumbertype', 'gidNumber')
+    ldapi_socket_raw = topo.standalone.config.get_attr_val_utf8('nsslapd-ldapifilepath')
+    ldapi_socket = ldapi_socket_raw.replace('/', '%2F')
+
+    # Create LDAP user
+    users = UserAccounts(topo.standalone, DEFAULT_SUFFIX)
+    user_properties = {
+        'uid': 'test_ldapi',
+        'cn': 'test_ldapi',
+        'sn': 'test_ldapi',
+        'uidNumber': '2020',
+        'gidNumber': '2020',
+        'userpassword': 'password',
+        'description': 'userdesc',
+        'homeDirectory': '/home/test_ldapi'}
+    users.create(properties=user_properties)
+
+    # Create OS user
+    subprocess.run(['useradd', '-p', LINUX_PWD, LINUX_USER ])
+
+    # Get UID GID
+    uid = getpwnam(LINUX_USER)[2]
+    gid = grp.getgrnam(LINUX_USER)[2]
+
+    # Create subtree to store LDAPI mappings under cn=config
+    config = DSLdapObject(topo.standalone, dn="cn=auto_bind,cn=config")
+    config._create_objectclasses = ['top', 'extensibleObject']
+    config.create(rdn="cn",
+                  basedn="cn=config",
+                  properties={'cn': 'auto_bind'})
+
+    # Create mapping entry
+    config = DSLdapObject(topo.standalone, dn="cn=entry_map1,cn=auto_bind,cn=config")
+    config._create_objectclasses = ['top', 'extensibleObject']
+    config.create(rdn="cn",
+                  basedn="cn=auto_bind,cn=config",
+                  properties={'cn': 'entry_map1',
+                              'uidNumber': str(uid),
```

Comment:
This may have issues if you run the test under root depending on the order of mapping precedence?

Reply:
Actually the test requires you run it as root :-)  Anyway I don't see a need for mapping precedence, since the server is doing an internal search that should only return a single entry.  If I am missing your point, please elaborate...

Reply:
Well I think my point is more that in the test it "breaks" root ldapi, but I think our test env doesn't actually rely on ldapi for root as we use the DM pw and ldap via localhost, so it's probably actually a non-concern.